keps/sig-network/3458-remove-transient-node-predicates-from-service-controller/README.md

```diff
@@ -0,0 +1,337 @@
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The service controller in the Kubernetes cloud controller manager (KCCM)
+currently adds/removes Nodes from the load balancers' node set in the following
+cases:
+
+a) When a node gets the taint `ToBeDeletedByClusterAutoscaler` added/removed
+b) When a node goes `Ready` / `NotReady`
+
+b) however only applies to services with `externalTrafficPolicy: Cluster`. In
+both cases: removing the Node in question from the load balancers' node set will
+cause all connections on that node to get terminated instantly. This can be
+considered a bug / sub-optimal behavior for nodes which are experiencing
+transient readiness state or for terminating nodes, since connections are not
+allowed to drain in those cases, even though the load balancer might support
+that. Moreover: on large clusters with a lot nodes and entropy, re-syncing load
+balancers like this can lead to rate-limiting by the cloud provider due to an
+excessive amount of update calls.
+
+As to enable connection draining, reduce cloud provider API calls and simplify
+the KCCMs sync loop: this KEP proposes that the service controller stops
+synchronizing the load balancer node set in these cases. Seeing as how this has
+always been the case, a new feature gate `LoadBalancerConnectionDraining` will
+be introduced, which will be used to enable the more optimal behavior.
+
+## Motivation
+
+Abruptly terminating connections in the cases defined by a) and b) above can be
+seen as buggy behavior and should be improved. By enabling connection draining,
+applications are allowed profit from graceful shutdown / termination, for what
+concerns cluster ingress connectivity. Users of Kubernetes will also see a
+reduction in the amount of cloud API calls, for what concerns calls stemming
+from syncing load balancers with the Kubernetes cluster state.
+
+### Goals
+
+- Stop re-configuring the load balancers' node set for cases a) and b) above
+
+### Non-Goals
+
+- Stop re-configuring the load balancers' node set for fully deleted /
+  newly added cluster nodes, or for nodes which get annotated with
+  `node.kubernetes.io/exclude-from-external-load-balancers`.
+
+## Proposal
+
+### Risks and Mitigations
+
+#### Risk
```

Comment:
What about the problem that still new connections might be coming to not-ready nodes (especially in the case if the non-readiness is not transient)?

Reply:
I mentioned this on the bigger KEP that we've decided to split it. What really matters for the load balancer is: "is kube-proxy healthy?". If the Node is `NotReady`, but kube-proxy healthy and able to serve responses to the load balancers' HC probes: then we should really continue using the Node for ingress traffic. And this remains untouched. What we're currently doing is worse, because: all Node conditions which transition the Node into `NotReady` do not have any correlation to the Node's networking capability

Reply:
I agree with this - my question is rather: "does all load-balancer for every provider use kube-proxy HC probes"?
If yes - then that's fine.
If not - then we would have a regression for those providers as they would have new connections coming to unhealthy nodes that wasn't happening before.

Reply:
> "does all load-balancer for every provider use kube-proxy HC probes"?

It depends. I am only aware of what the "big three" do:

- GCP: probes port 10256 (Kube-proxy's healthz port)
- AWS: if ELB; probes the first `NodePort` defined on the service spec
- Azure: probes all `NodePort` defined on the service spec.

All clouds take an approach of trying to ascertain if traffic can be forwarded to the endpoint. In GCP's case it checks if kube-proxy (or service proxy, rather) is healthy, while the others validate that the service proxy has accomplished what it needs for the service/load balancer in question. 

But I see your point, and I agree: we are not sure _all_ cloud providers do, maybe some do it completely wrong and check something else? Given that we'll be using a feature gate for this: they could opt-out should they do something else (not very likely). I will add this to the Risks.

Reply:
If the network plugin on a node goes down, and stops being able to route traffic to pods, then the node will (at least potentially) become un-Ready, but kube-proxy is not guaranteed to also become unhealthy. So in that case, the AWS/Azure style (of probing nodeports) would still work, but the GCP style might not.

I had never realized that GCP and AWS/Azure did things differently here. It sounds like we don't have a very good definition of how cloud LBs are supposed to be checking the readiness of (Cluster traffic policy) services (or inversely, how kubernetes is supposed to be indicating the readiness of services to cloud load balancers).

Reply:
> I had never realized that GCP and AWS/Azure did things differently here.

Interesting. Should figuring this our be a prerequisite for this KEP then? 
@thockin

Reply:
I don't think so.  All of these modes have drawbacks.  Dan cite's GCP's.

Probing directly to NodePort just proves that at least one backend exists, not that the dataplane has been updated recently.  Worse, a single failing backend pod *on a different node* can take a whole node out of the LB path (which is why GCP *didn't* do it that way.

IMO the provider needs to make this decision with knowledge of how their LB works.  For GCP, the LB knows about the VM and wants to prove that the VM is up and reasonably up-to-date.  It might be interesting to ALSO prove that the dataplane is routing, but we can't give a single pod the ability to fail out a whole node.

For a provider which routes directly to pods, the node health is the wrong thing to probe.

> they would have new connections coming to unhealthy nodes that wasn't happening before

That's true, but given the under-definedness of node readiness, I don't think it actually means anything useful.

Reply:
Thanks Tim - I'm sold on that.
That said, I think it would be useful to add this context to the KEP itself for future readers (and I bet I may come up with exactly the same question next release, so having a clear answer for that in the KEP would also help future me :D )